[Add blog post]  add Compat2021 end of year blog post (#7011)
GoogleChrome/web.dev: src/site/content/en/blog/compat2021-holiday-update/index.md

---
title: "Compat 2021 Holiday Update: presents for developers before the end of the year"
subhead:
    "End of the year update on Compat 2021—an effort to eliminate browser compatibility problems in five key focus areas: CSS Flexbox, CSS Grid, position: sticky, aspect-ratio, and CSS transforms."
authors:
  - foolip
  - kosamari 
hero: "image/Wd2sVnt4VLho4jgp7UtIyWFceE02/T4gzpGXXQ9sud6CEbHno.jpg" alt: "a hand holding a small box of presents with red ribbon." 

> date: 2021-12-10 
> updated: 2021-12-10 

Reviewer: ```suggestion
date: 2021-12-13
```

tags:
  - blog
  - css
---

The end of the year is near, and it's time for a final update on Compat 2021—an effort to eliminate 
browser compatibility problems in five key focus areas.

{% Aside %}
This effort is also known as Interop 2021, matching the name of 
[Interop 2022](https://github.com/web-platform-tests/interop-2022), which is the continuation of 
this effort.
{% endAside %}

<div class="w-stats">
 <div class="w-stat">
   <p class="w-stat__figure">>90<sub>%</sub></p>
   <p class="w-stat__desc">score in all browsers</p>
 </div>
</div>

Since our [last update](https://web.dev/compat2021-midyear/), we’ve continued to see improvements 
in all browsers. All browsers started with much lower test scores at the beginning of the year, but 
now all browsers have surpassed 90%! This means the web platform has significantly improved 

> interoperability of 5 focus areas. 

Reviewer 2: ```suggestion
interoperability of the 5 focus areas. 
```

<figure class="w-figure">
{% Img src="image/Wd2sVnt4VLho4jgp7UtIyWFceE02/Kaz3ye7gFfow8ia7lWYW.png", alt="A snapshot of Compat 
2021 Dashboard (experimental browsers)", width="800", height="727" %}
  <figcaption class="w-figcaption">
    A snapshot of <a href="https://wpt.fyi/compat2021">Compat 2021 Dashboard</a> (experimental 
    browsers).
  </figcaption>
</figure>

{% Aside 'caution' %}
The graph does not correctly show the improvements in Safari Technology Preview between May and 
November, as the version tested was [stuck](https://github.com/web-platform-tests/wpt/issues/31147) 
at an older version. This is now resolved but the data between May and November remains incorrect.
{% endAside %}

Contributions to browser engines are made not only by browser vendors, but also others in the web 
community. For this project, we particularly want to thank Igalia for 
[their involvement](https://www.igalia.com/2021/11/12/New-Interoperability-Milestones.html) and 

> their continued work to improve the scores. Igalia has contributed to improving all five focus 

Reviewer 2: ```suggestion
continued work to improve the scores. Igalia has contributed to improving all five focus 
```

areas of Compat 2021.

On [wpt.fyi](https://wpt.fyi/), the test results dashboard, there’s now a 
[filtered test results view](https://wpt.fyi/results/?label=master&label=experimental&product=chrome&product=firefox&product=safari&aligned&q=label%3Ainterop-2021)

> ([permalink](https://wpt.fyi/results/?q=label%3Ainterop-2021&run_id=5629744973348864&run_id=5117500464300032&run_id=5734588178497536)) showing all of the tests included in Compat 2021, 

Reviewer 3: from the context it's not totally clear to me that this means a permalink of both the test filtering _and_ the specific runs; maybe be explicit, or just expand it to "(as of today)" or something?

Reviewer 4: I'll remove it. The 3 runs that were current can be dug out of the 3 following links, and I'm pretty sure I'm the only person who ever might.

> and also views for [Chrome](https://wpt.fyi/results/?diff&filter=ADC&q=label%3Ainterop-2021&run_id=5738932147847168&run_id=5629744973348864), 

Reviewer 2: ```suggestion
and also views for [Chrome/Edge](https://wpt.fyi/results/?diff&filter=ADC&q=label%3Ainterop-2021&run_id=5738932147847168&run_id=5629744973348864), 
```

[Firefox](https://wpt.fyi/results/?diff&filter=ADC&q=label%3Ainterop-2021&run_id=5644039006191616&run_id=5117500464300032), 
and [Safari](https://wpt.fyi/results/?diff&filter=ADC&q=label%3Ainterop-2021&run_id=5739124314079232&run_id=5734588178497536) comparing the results to our last update in July.

Let’s take a look at the improvements in each area!

## CSS flexbox

`flex-basis: content` is now on its way to all browsers, with implementations landing in 
[Chromium](https://bugs.chromium.org/p/chromium/issues/detail?id=470421) and 
[WebKit](https://trac.webkit.org/changeset/284440/webkit). (The `content` value was already 

> supported by Firefox.)

Reviewer 2: ```suggestion
supported by Gecko.)
```

Reviewer 2: We should consistently use brand or repo names, and also help explain to those who don't know what the words mean.

Reviewer 3: Does this not also imply that we should also be using Blink instead of Chromium?

Reviewer 4: We used Chromium as the engine name in both https://web.dev/compat2021/ and https://web.dev/compat2021-midyear/.

Reviewer 4: I should also say I think that's not a bug, Chromium is the name of a browser engine, and Blink is a part of Chromium. So when do we call it Blink instead of Chromium? I don't know :)

In Chromium, an [issue with flexbox sizing](https://bugs.chromium.org/p/chromium/issues/detail?id=961902) 

> is fixed, matching the spec and Firefox’s behavior. And in Gecko, several 

Reviewer 2: ```suggestion
is fixed, matching the spec and Gecko's behavior. And in Gecko, several 
```

[issues affecting Compat 2021](https://bugzilla.mozilla.org/show_bug.cgi?id=1700745) are fixed, 
including an [issue with percentage height on flex items](https://bugzilla.mozilla.org/show_bug.cgi?id=1611303). 

> Finally, in WebKit, support for more alignment property values ([left, right](https://trac.webkit.org/changeset/282078/webkit)
> , [self-start, self-end](https://trac.webkit.org/changeset/282267/webkit), [start, end](https://trac.webkit.org/changeset/281840/webkit)) 

Reviewer 4: ```suggestion
Finally, in WebKit, support for more alignment property values ([left, right](https://trac.webkit.org/changeset/282078/webkit),
[self-start, self-end](https://trac.webkit.org/changeset/282267/webkit), [start, end](https://trac.webkit.org/changeset/281840/webkit)) 
```

is now added, and a lot of improvements were made for [absolute positioning](https://trac.webkit.org/changeset/281995/webkit), 
also improving the flexbox test results in Compat 2021.

## CSS Grid

The use of CSS Grid on the web continues to grow, as can be seen in both the 
[2021 Web Almanac](https://almanac.httparchive.org/en/2021/css#flexbox-and-grid-adoption) and 
Chrome’s [usage metrics](https://www.chromestatus.com/metrics/feature/timeline/popularity/1693).

The [launch of GridNG](https://blogs.windows.com/msedgedev/2021/08/10/compat2021-css-grid-gridng/) 
in Chrome and Edge 93 resolved many long standing issues with Grid, closing an impressive 38 issues 
in Chromium’s bug tracker. Together with many smaller improvements that followed, the Compat 2021 
score for Grid in Chromium improved by 3% to 97%. This work was led by the Edge team at Microsoft.

An [absolute positioning bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1707643) affecting Grid 

> was fixed in Gecko, and [many fixes](https://bugs.webkit.org/buglist.cgi?bug_status=RESOLVED&email1=zsun%40igalia.com&emailassigned_to1=1&emailtype1=equals&list_id=7730965&query_format=advanced&resolution=FIXED&short_desc=grid&short_desc_type=allwordssubstr)

Reviewer 3: Maybe instead https://bugs.webkit.org/buglist.cgi?bug_status=RESOLVED&chfield=resolution&chfieldfrom=2021-01-01&chfieldto=2021-12-31&component=Accessibility&component=CSS&component=Layout%20and%20Rendering&component=New%20Bugs&f1=short_desc&f2=short_desc&f3=short_desc&list_id=7736912&o1=notsubstring&o2=notsubstring&o3=substring&query_format=advanced&resolution=FIXED&v1=Web%20Inspector&v2=%5BLFC%5D&v3=grid (which has quite a lot more fixes!)

Reviewer 4: Thanks, let's go with that!

 have landed in WebKit, leading to a 1% improvement for Firefox and 3% improvement for Safari on 
the Grid tests.

## CSS `position: sticky`

In our last update, we noted that `position: sticky` was the first area where any browser (in this 

> case Chrome and Edge) reached 100% passing tests. Now, following a [number of fixes](https://bugs.webkit.org/buglist.cgi?bug_status=RESOLVED&email1=mrobinson%40webkit.org&emailassigned_to1=1&emailtype1=equals&list_id=7731158&query_format=advanced&resolution=FIXED&short_desc=sticky&short_desc_type=allwordssubstr) 

Reviewer 3: https://bugs.webkit.org/buglist.cgi?bug_status=RESOLVED&chfield=resolution&chfieldfrom=2021-01-01&chfieldto=2021-12-31&f1=short_desc&f2=short_desc&f3=short_desc&list_id=7736931&o1=notsubstring&o2=notsubstring&o3=substring&query_format=advanced&resolution=FIXED&v1=Web%20Inspector&v2=%5BLFC%5D&v3=sticky again has more?

Reviewer 4: 👍

in WebKit’s implementation, Safari also scores 100% for these tests.  Most of these improvements 
were included in Safari 15.

## CSS `aspect-ratio` property

Cross-browser support for defining the aspect ratio (width-to-height ratio) of elements has 
continued to improve, with Compat 2021 scores reaching 99%, 97% and 95% for Chrome/Edge, Firefox 
and Safari respectively. Most of the improvements are not with the `aspect-ratio` property itself, 
but rather with how `width` and `height` attributes are [mapped to a default `aspect-ratio` value](https://developer.mozilla.org/docs/Web/Media/images/aspect_ratio_mapping) 
for elements. This was implemented for multiple elements in [WebKit](https://wpt.fyi/results/html/rendering/replaced-elements/attributes-for-embedded-content-and-images?diff&filter=ADC&q=label%3Ainterop-2021&run_id=5739124314079232&run_id=6290692121821184), 
and `<canvas>` for [Chromium](https://chromium-review.googlesource.com/c/chromium/src/+/3109968).

## CSS transforms

Support for `transform: perspective(none)` is now supported in 
[Chromium](https://chromestatus.com/feature/5687325523705856), 
[Gecko](https://bugzilla.mozilla.org/show_bug.cgi?id=1725207) and 

> [WebKit](https://trac.webkit.org/changeset/285255/webkit). This will make it easier to do 

Reviewer 2: ```suggestion
[WebKit](https://trac.webkit.org/changeset/285255/webkit). This will make it easier to 
```

> animations between a perspective and no perspective.

Reviewer 2: ```suggestion
animate between a perspective and no perspective.
```

In Chromium, `transform-style: preserve-3d` (which allows child elements to participate in the same 
3D scene) and the `perspective` property (which applies a perspective transform to child elements) 
are now [aligned with the spec](https://chromestatus.com/feature/5640541339385856) by making them 
apply only to child elements.

The big increase in the [scores](https://wpt.fyi/compat2021?feature=css-transforms) for CSS 

> transforms for all browsers is mainly due to improvements to the test suite, where some incorrect 

Reviewer 2: ```suggestion
transforms for all browsers is mainly due to improvements to the test suite, where incorrect 
```

tests have been fixed or removed. This makes it easier to understand the remaining interoperability 

> problems, and avoids regressions in the future.

Reviewer 2: ```suggestion
problems and avoid regressions in the future.
```

## Conclusion

We are grateful for the work that everyone has put in to end the year with many improvements to the 
score as well as better testing infrastructure. `aspect-ratio` was a long requested feature from 

> web developers and it is now supported in all browsers. Growth in use of flexbox and grid layout is 

Reviewer 2: The sentence starting with "Growth" doesn't quite parse.

Reviewer 4: Agreed. @kosamari I'll reword it as:

> And the growing use flexbox and grid layout is now better supported thanks to many improvements made during 2021.

Reviewer 4: Hmm, it's not the *growth* that's better supported, it's the features themselves. Instead:

> Use of flexbox, grid and `position: sticky` are all growing, and these features are now better supported across browsers thanks to many improvements made during 2021.

Added `position: sticky` after looking at https://chromestatus.com/metrics/feature/timeline/popularity/3354

---

> title: "Compat2021 Holiday Update: presents for developers before the end of the year"

Reviewer 4: ```suggestion
title: "Compat 2021 Holiday Update: presents for developers before the end of the year"
```

subhead:

>     "End of the year update on Compat2021 —an effort to eliminate browser compatibility problems in five key focus areas: CSS Flexbox, CSS Grid, position: sticky, aspect-ratio, and CSS transforms."

Reviewer 4: ```suggestion
    "End of the year update on Compat 2021—an effort to eliminate browser compatibility problems in five key focus areas: CSS Flexbox, CSS Grid, position: sticky, aspect-ratio, and CSS transforms."
```

this effort.
{% endAside %}

<div class="w-stats">
 <div class="w-stat">
   <p class="w-stat__figure">>90<sub>%</sub></p>
   <p class="w-stat__desc">score in all browsers</p>
 </div>
</div>

Since our [last update](https://web.dev/compat2021-midyear/), we’ve continued to see improvements 
in all browsers. All browsers started with much lower test scores at the beginning of the year, but 
now all browsers have surpassed 90%! This means the web platform has significantly improved 
interoperability of 5 focus areas. 

<figure class="w-figure">
{% Img src="image/Wd2sVnt4VLho4jgp7UtIyWFceE02/Kaz3ye7gFfow8ia7lWYW.png", alt="A snapshot of Compat 
2021 Dashboard (experimental browsers)", width="800", height="727" %}
  <figcaption class="w-figcaption">
    A snapshot of <a href="https://wpt.fyi/compat2021">Compat 2021 Dashboard</a> (experimental 
    browsers).
  </figcaption>
</figure>

{% Aside 'caution' %}
The graph does not correctly show the improvements in Safari Technology Preview between May and 
November, as the version tested was [stuck](https://github.com/web-platform-tests/wpt/issues/31147) 

> at older versions. This is now resolved but the data between May and November remains incorrect.

Reviewer 4: ```suggestion
at an older version. This is now resolved but the data between May and November remains incorrect.
```

and also views for [Chrome](https://wpt.fyi/results/?diff&filter=ADC&q=label%3Ainterop-2021&run_id=5738932147847168&run_id=5629744973348864)
, [Firefox](https://wpt.fyi/results/?diff&filter=ADC&q=label%3Ainterop-2021&run_id=5644039006191616&run_id=5117500464300032)
, and [Safari](https://wpt.fyi/results/?diff&filter=ADC&q=label%3Ainterop-2021&run_id=5739124314079232&run_id=5734588178497536) 
comparing the results to our last update in July.

Let’s take a look at the improvements in each area!

## CSS flexbox

`flex-basis: content` is now on its way to all browsers, with implementations landing in 
[Chromium](https://bugs.chromium.org/p/chromium/issues/detail?id=470421) and 
[WebKit](https://trac.webkit.org/changeset/284440/webkit). (The `content` value was already 
supported by Firefox.)

In Chromium, an [issue with flexbox sizing](https://bugs.chromium.org/p/chromium/issues/detail?id=961902) 
is fixed, matching the spec and Firefox’s behavior. And in Gecko, several 
[issues affecting Compat 2021](https://bugzilla.mozilla.org/show_bug.cgi?id=1700745) are fixed, 

> including an [issue with percentage height on flex items](https://bugzilla.mozilla.org/show_bug.cgi?id=1611303)
> . Finally, in WebKit, support for more alignment property values ([left, right](https://trac.webkit.org/changeset/282078/webkit)

Reviewer 4: There's a space before the period in the staging instance. Fix:

```suggestion
including an [issue with percentage height on flex items](https://bugzilla.mozilla.org/show_bug.cgi?id=1611303).
Finally, in WebKit, support for more alignment property values ([left, right](https://trac.webkit.org/changeset/282078/webkit)
```

> , [self-start, self-end](https://trac.webkit.org/changeset/282267/webkit), [start, end](https://trac.webkit.org/changeset/281840/webkit)
> ) is now added, and a lot of improvements were made for [absolute positioning](https://trac.webkit.org/changeset/281995/webkit)

Reviewer 4: Space before ) here:

```suggestion
, [self-start, self-end](https://trac.webkit.org/changeset/282267/webkit), [start, end](https://trac.webkit.org/changeset/281840/webkit))
is now added, and a lot of improvements were made for [absolute positioning](https://trac.webkit.org/changeset/281995/webkit)
```

 have landed in WebKit, leading to a 1% improvement for Firefox and 3% improvement for Safari on 
the Grid tests.

## CSS `position: sticky`

In our last update, we noted that `position: sticky` was the first area where any browser (in this 
case Chrome and Edge) reached 100% passing tests. Now, following a [number of fixes](https://bugs.webkit.org/buglist.cgi?bug_status=RESOLVED&email1=mrobinson%40webkit.org&emailassigned_to1=1&emailtype1=equals&list_id=7731158&query_format=advanced&resolution=FIXED&short_desc=sticky&short_desc_type=allwordssubstr) 
in WebKit’s implementation, Safari also scores 100% for these tests.  Most of these improvements 
were included in Safari 15.

## CSS `aspect-ratio` property

Cross-browser support for defining the aspect ratio (width-to-height ratio) of elements has 
continued to improve, with Compat 2021 scores reaching 99%, 97% and 95% for Chrome/Edge, Firefox 
and Safari respectively. Most of the improvements are not with the `aspect-ratio` property itself, 
but rather with how `width` and `height` attributes are [mapped to a default `aspect-ratio` value](https://developer.mozilla.org/docs/Web/Media/images/aspect_ratio_mapping) 

> for elements. This was implemented for multiple elements in [WebKit](https://wpt.fyi/results/html/rendering/replaced-elements/attributes-for-embedded-content-and-images?diff&filter=ADC&q=label%3Ainterop-2021&run_id=5739124314079232&run_id=6290692121821184)
> , and `&lt;canvas>` for [Chromium](https://chromium-review.googlesource.com/c/chromium/src/+/3109968).

Reviewer 4: ```suggestion
for elements. This was implemented for multiple elements in [WebKit](https://wpt.fyi/results/html/rendering/replaced-elements/attributes-for-embedded-content-and-images?diff&filter=ADC&q=label%3Ainterop-2021&run_id=5739124314079232&run_id=6290692121821184),
and `<canvas>` for [Chromium](https://chromium-review.googlesource.com/c/chromium/src/+/3109968).
```

---
title: "Compat 2021 Holiday Update: presents for developers before the end of the year"
subhead:
    "End of the year update on Compat 2021—an effort to eliminate browser compatibility problems in five key focus areas: CSS Flexbox, CSS Grid, position: sticky, aspect-ratio, and CSS transforms."
authors:
  - foolip
  - kosamari

> hero: "image/Wd2sVnt4VLho4jgp7UtIyWFceE02/T4gzpGXXQ9sud6CEbHno.jpg" alt: "a hand holding a small box of presents with red ribbon."

Reviewer: ```suggestion
hero: "image/Wd2sVnt4VLho4jgp7UtIyWFceE02/T4gzpGXXQ9sud6CEbHno.jpg" 
alt: "a hand holding a small box of presents with red ribbon."
```

date: 2021-12-13
tags:
  - blog
  - css
---

The end of the year is near, and it's time for a final update on Compat 2021—an effort to eliminate
browser compatibility problems in five key focus areas.

{% Aside %}
This effort is also known as Interop 2021, matching the name of
[Interop 2022](https://github.com/web-platform-tests/interop-2022), which is the continuation of
this effort.
{% endAside %}

<div class="w-stats">
 <div class="w-stat">
   <p class="w-stat__figure">>90<sub>%</sub></p>
   <p class="w-stat__desc">score in all browsers</p>
 </div>
</div>

> Since our [last update](https://web.dev/compat2021-midyear/), we’ve continued to see improvements

Reviewer: ```suggestion
Since our [last update](/compat2021-midyear/), we’ve continued to see improvements
```

in all browsers. All browsers started with much lower test scores at the beginning of the year, but
now all browsers have surpassed 90%! This means the web platform has significantly improved
interoperability of the five focus areas.

> <figure class="w-figure">
> {% Img src="image/Wd2sVnt4VLho4jgp7UtIyWFceE02/Kaz3ye7gFfow8ia7lWYW.png", alt="A snapshot of Compat
> 2021 Dashboard (experimental browsers)", width="800", height="727" %}
>   <figcaption class="w-figcaption">
>     A snapshot of <a href="https://wpt.fyi/compat2021">Compat 2021 Dashboard</a> (experimental
>     browsers).
>   </figcaption>
> </figure>

Reviewer: ```suggestion
<figure>
{% Img src="image/Wd2sVnt4VLho4jgp7UtIyWFceE02/Kaz3ye7gFfow8ia7lWYW.png", alt="A snapshot of Compat
2021 Dashboard (experimental browsers)", width="800", height="727" %}
  <figcaption>
    A snapshot of <a href="https://wpt.fyi/compat2021">Compat 2021 Dashboard</a> (experimental
    browsers).
  </figcaption>
</figure>
```
